File: 백준/Silver/1743. 음식물 피하기/음식물 피하기.py
import sys
input = sys.stdin.readline
print = sys.stdout.write

def function():
    # write code down

    # input
    # 1. m, n, k : 세로, 가로, 음식물 쓰레기 개수 (1 <= n,m <= 100, 1 <= k <= n*m)
    # 2~. y, x : 좌표

    # output
    # 영역이 가장 큰 음식물을 출력
    global visited, res, dx, dy, n, m

    m, n, k = map(int, input().split())
    graph = [[0]*n for _ in range(m)]
    for _ in range(k):
        y, x = map(int, input().split())
        graph[y-1][x-1] = 1
    visited = [[False]*n for _ in range(m)]

    res = 0

    dx = [0,0,1,-1]
    dy = [1,-1,0,0]

    for i in range(m):
        for j in range(n):
            if graph[i][j]==1 and not visited[i][j]:
                visited[i][j] = True
                bfs(graph, (j, i))
    
    
    return str(res)

def bfs(graph, start):
    global visited, res, dx, dy, n, m
    
    from collections import deque

    q = deque()
    q.append(start)
    cnt = 0

    while q:
        x, y = q.popleft()
        cnt += 1

        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]

            if 0 <= nx < n and 0 <= ny < m and graph[ny][nx]==1 and not visited[ny][nx]:
                q.append((nx, ny))
                visited[ny][nx] = True

    res = max(res, cnt)

if __name__ == '__main__':
    print(function())

File: 백준/Silver/1743. 음식물 피하기/음식물 피하기.java
import java.io.*;
import java.util.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static void read() throws IOException{
        st = new StringTokenizer(br.readLine());
    }
    static int nextInt() {
        return Integer.parseInt(st.nextToken());
    }

    public static class Point {
        int y;
        int x;
        Point(int a, int b){
            y = a;
            x = b;
        }
    }

    static int n, m, k, res;
    static int[][] graph;
    static boolean[][] visited;

    static int[] dy = {1,-1,0,0};
    static int[] dx = {0,0,1,-1};

    static void bfs(Point start){
        Deque<Point> q = new LinkedList<>();
        visited[start.y][start.x] = true;
        q.addFirst(start);

        int cnt = 0;

        while(!q.isEmpty()){
            Point node = q.pollLast();
            cnt++;

            for(int i=0; i<4; i++){
                int ny = node.y + dy[i];
                int nx = node.x + dx[i];

                if(0 <= ny && ny < m && 0 <= nx && nx < n && graph[ny][nx]==1 && !visited[ny][nx]){
                    q.addFirst(new Point(ny, nx));
                    visited[ny][nx] = true;
                }
            }
        }

        res = Math.max(res, cnt);
    }
    
    public static void main(String[] args) throws IOException{
        read();
        m = nextInt();
        n = nextInt();
        k = nextInt();
        graph = new int[m][n];
        visited = new boolean[m][n];

        for(int i=0; i<k; i++){
            read();
            graph[nextInt()-1][nextInt()-1] = 1;
        }

        res = 0;

        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(graph[i][j]==1 && !visited[i][j]){
                    bfs(new Point(i, j));
                }
            }
        }

        System.out.print(res);
        
    }
}